python_modules/dagster/dagster/_core/execution/context/system.py

```diff
@@ -975,26 +975,26 @@ def partition_mapping_for_input(self, input_name: str) -> Optional[PartitionMapp
         else:
             return None
 
-    def _get_input_asset_event(self, key: AssetKey, retries: int = 0) -> Optional["EventLogRecord"]:
-        from dagster._core.definitions.data_version import (
-            extract_data_version_from_entry,
-        )
+    def _get_input_asset_event(self, key: AssetKey) -> Optional["EventLogRecord"]:
+        from dagster._core.events import DagsterEventType
 
-        is_step_internal_input = key in self._data_version_cache
-        event = self.instance.get_latest_data_version_record(key)
-        if event is None and is_step_internal_input and retries <= 5:
-            return self._get_input_asset_event(key, retries + 1)
-        elif event is None:
-            return None
+        # step-internal input
+        if key in self._data_version_cache:  # cannot be an observation
+            cursor = None
+            records = []
+            while True:
+                connection = self.instance.get_records_for_run(
+                    self.run_id, of_type={DagsterEventType.ASSET_MATERIALIZATION}, cursor=cursor
+                )
+                records.extend(connection.records)
+                cursor = connection.cursor
+                if not connection.has_more:
+                    break
```

Comment:
Is there a helper funciton already in the codebase for this? (getting _all_ run records matching a query instead of a connection)